From d9c0f98127938abcfb073cfc8d75645b90548a53 Mon Sep 17 00:00:00 2001
From: Gabriel de Araújo Cabral <gabriel.cabral@windriver.com>
Date: Thu, 18 Apr 2024 08:36:28 -0300
Subject: [PATCH] ceph-csi-rbd: update storageclass

This patch updates some fields and values for the rbd storage class.

Signed-off-by: Gabriel de Araújo Cabral <gabriel.cabral@windriver.com>
---
 .../ceph-csi-rbd/templates/storageclass.yaml  | 198 ++++++++++--------
 1 file changed, 115 insertions(+), 83 deletions(-)

diff --git a/charts/ceph-csi-rbd/templates/storageclass.yaml b/charts/ceph-csi-rbd/templates/storageclass.yaml
index 188362fc0..d971ed1fe 100644
--- a/charts/ceph-csi-rbd/templates/storageclass.yaml
+++ b/charts/ceph-csi-rbd/templates/storageclass.yaml
@@ -1,96 +1,128 @@
 {{- if .Values.storageClass.create -}}
+{{- $cephCsiDefault := .Values.storageClass }}
+{{- $defaults := .Values.classdefaults }}
+{{- range $sc := .Values.storageClasses }}
 apiVersion: storage.k8s.io/v1
 kind: StorageClass
 metadata:
-  name: {{ .Values.storageClass.name }}
-{{- if .Values.storageClass.annotations }}
+  name: {{ $sc.name }}
   annotations:
-{{ toYaml .Values.storageClass.annotations | indent 4 }}
-{{- end }}
+  {{- if eq $defaults.storageClass $sc.name }}
+    "storageclass.kubernetes.io/is-default-class": "true"
+  {{- end }}
+{{ toYaml $cephCsiDefault.annotations | indent 4 }}
   labels:
-    app: {{ include "ceph-csi-rbd.name" . }}
-    chart: {{ include "ceph-csi-rbd.chart" . }}
-    release: {{ .Release.Name }}
-    heritage: {{ .Release.Service }}
-    {{- with .Values.commonLabels }}{{ toYaml . | trim | nindent 4 }}{{- end }}
-provisioner: {{ .Values.driverName }}
+    app: {{ include "ceph-csi-rbd.name" $ }}
+    chart: {{ include "ceph-csi-rbd.chart" $ }}
+    release: {{ $.Release.Name }}
+    heritage: {{ $.Release.Service }}
+    {{- with $.Values.commonLabels }}{{ toYaml . | trim | nindent 4 }}{{- end }}
+provisioner: {{ $.Values.driverName }}
 parameters:
-  clusterID: {{ .Values.storageClass.clusterID }}
-  imageFeatures: {{ .Values.storageClass.imageFeatures }}
-{{- if .Values.storageClass.pool }}  
-  pool: {{ .Values.storageClass.pool }}
-{{- end }}  
-{{- if .Values.storageClass.tryOtherMounters }}
-  tryOtherMounters: {{ .Values.storageClass.tryOtherMounters | quote}}
-{{- end }}
-{{- if .Values.storageClass.mkfsOptions }}
-  mkfsOptions: {{ .Values.storageClass.mkfsOptions }}
-{{- end }}
-{{- if .Values.storageClass.mounter }}
-  mounter: {{ .Values.storageClass.mounter }}
-{{- end }}
-{{- if .Values.storageClass.cephLogDir }}
-  cephLogDir: {{ .Values.storageClass.cephLogDir }}
-{{- end }}
-{{- if .Values.storageClass.cephLogStrategy }}
-  cephLogStrategy: {{ .Values.storageClass.cephLogStrategy }}
-{{- end }}
-{{- if .Values.storageClass.dataPool }}
-  dataPool: {{ .Values.storageClass.dataPool }}
-{{- end }}
-{{- if .Values.storageClass.volumeNamePrefix }}
-  volumeNamePrefix: "{{ .Values.storageClass.volumeNamePrefix }}"
-{{- end }}
-{{- if .Values.storageClass.encrypted }}
-  encrypted: "{{ .Values.storageClass.encrypted }}"
-{{- end }}
-{{- if .Values.storageClass.encryptionKMSID }}
-  encryptionKMSID: {{ .Values.storageClass.encryptionKMSID }}
-{{- end }}
-{{- if .Values.storageClass.topologyConstrainedPools }}
+{{- with (first $.Values.csiConfig) }}
+  clusterID: {{ or $sc.clusterID .clusterID }}
+{{- end}}
+{{- if $sc.pool_name }}
+  pool: {{ $sc.pool_name }}
+{{- end }}
+  imageFeatures: {{ or $sc.imageFeatures $cephCsiDefault.imageFeatures }}
+{{- if $sc.tryOtherMounters }}
+  tryOtherMounters: {{ $sc.tryOtherMounters | quote}}
+{{- else if $cephCsiDefault.tryOtherMounters }}
+  tryOtherMounters: {{ $cephCsiDefault.tryOtherMounters | quote}}
+{{- end }}
+{{- if $sc.mkfsOptions }}
+  mkfsOptions: {{ $sc.mkfsOptions }}
+{{- else if $cephCsiDefault.mkfsOptions }}
+  mkfsOptions: {{ $cephCsiDefault.mkfsOptions }}
+{{- end }}
+{{- if $sc.mounter }}
+  mounter: {{ $sc.mounter }}
+{{- else if $cephCsiDefault.mounter }}
+  mounter: {{ $cephCsiDefault.mounter }}
+{{- end }}
+{{- if $sc.cephLogDir }}
+  cephLogDir: {{ $sc.cephLogDir }}
+{{- else if $cephCsiDefault.cephLogDir }}
+  cephLogDir: {{ $cephCsiDefault.cephLogDir }}
+{{- end }}
+{{- if $sc.cephLogStrategy }}
+  cephLogStrategy: {{ $sc.cephLogStrategy }}
+{{- else if $cephCsiDefault.cephLogStrategy }}
+  cephLogStrategy: {{ $cephCsiDefault.cephLogStrategy }}
+{{- end }}
+{{- if $sc.dataPool }}
+  dataPool: {{ $sc.dataPool }}
+{{- else if $cephCsiDefault.dataPool }}
+  dataPool: {{ $cephCsiDefault.dataPool }}
+{{- end }}
+{{- if $sc.volumeNamePrefix }}
+  volumeNamePrefix: "{{ $sc.volumeNamePrefix }}"
+{{- else if $cephCsiDefault.volumeNamePrefix }}
+  volumeNamePrefix: "{{ $cephCsiDefault.volumeNamePrefix }}"
+{{- end }}
+{{- if $sc.encrypted }}
+  encrypted: "{{ $sc.encrypted }}"
+{{- else if $cephCsiDefault.encrypted }}
+  encrypted: "{{ $cephCsiDefault.encrypted }}"
+{{- end }}
+{{- if $sc.encryptionKMSID }}
+  encryptionKMSID: {{ $sc.encryptionKMSID }}
+{{- else if $cephCsiDefault.encryptionKMSID }}
+  encryptionKMSID: {{ $cephCsiDefault.encryptionKMSID }}
+{{- end }}
+{{- if $sc.topologyConstrainedPools }}
   topologyConstrainedPools:
-  {{ toYaml .Values.storageClass.topologyConstrainedPools | indent 4 -}}
-{{- end }}
-{{- if .Values.storageClass.mapOptions }}
-  mapOptions: {{ .Values.storageClass.mapOptions }}
-{{- end }}
-{{- if .Values.storageClass.unmapOptions }}
-  unmapOptions: {{ .Values.storageClass.unmapOptions }}
-{{- end }}
-{{- if .Values.storageClass.stripeUnit }}
-  stripeUnit: {{ .Values.storageClass.stripeUnit | quote }}
-{{- end }}
-{{- if .Values.storageClass.stripeCount }}
-  stripeCount: {{ .Values.storageClass.stripeCount | quote }}
-{{- end }}
-{{- if .Values.storageClass.objectSize }}
-  objectSize: {{ .Values.storageClass.objectSize | quote }}
-{{- end }}
-  csi.storage.k8s.io/provisioner-secret-name: {{ .Values.storageClass.provisionerSecret }}
-{{- if .Values.storageClass.provisionerSecretNamespace }}
-  csi.storage.k8s.io/provisioner-secret-namespace: {{ .Values.storageClass.provisionerSecretNamespace }}
-{{ else }}
-  csi.storage.k8s.io/provisioner-secret-namespace: {{ .Release.Namespace }}
-{{- end }}
-  csi.storage.k8s.io/controller-expand-secret-name: {{ .Values.storageClass.controllerExpandSecret }}
-{{- if .Values.storageClass.controllerExpandSecretNamespace }}
-  csi.storage.k8s.io/controller-expand-secret-namespace: {{ .Values.storageClass.controllerExpandSecretNamespace }}
-{{ else }}
-  csi.storage.k8s.io/controller-expand-secret-namespace: {{ .Release.Namespace }}
-{{- end }}
-  csi.storage.k8s.io/node-stage-secret-name: {{ .Values.storageClass.nodeStageSecret }}
-{{- if .Values.storageClass.nodeStageSecretNamespace }}
-  csi.storage.k8s.io/node-stage-secret-namespace: {{ .Values.storageClass.nodeStageSecretNamespace }}
-{{ else }}
-  csi.storage.k8s.io/node-stage-secret-namespace: {{ .Release.Namespace }}
-{{- end }}
-  csi.storage.k8s.io/fstype: {{ .Values.storageClass.fstype }}
-reclaimPolicy: {{ .Values.storageClass.reclaimPolicy }}
-allowVolumeExpansion: {{ .Values.storageClass.allowVolumeExpansion }}
-{{- if .Values.storageClass.mountOptions }}
+  {{ toYaml $sc.topologyConstrainedPools | indent 4 -}}
+{{- else if $cephCsiDefault.topologyConstrainedPools }}
+  topologyConstrainedPools:
+  {{ toYaml $cephCsiDefault.topologyConstrainedPools | indent 4 -}}
+{{- end }}
+{{- if $sc.mapOptions }}
+  mapOptions: {{ $sc.mapOptions }}
+{{- else if $cephCsiDefault.mapOptions }}
+  mapOptions: {{ $cephCsiDefault.mapOptions }}
+{{- end }}
+{{- if $sc.unmapOptions }}
+  unmapOptions: {{ $sc.unmapOptions }}
+{{- else if $cephCsiDefault.unmapOptions }}
+  unmapOptions: {{ $cephCsiDefault.unmapOptions }}
+{{- end }}
+{{- if $sc.stripeUnit }}
+  stripeUnit: {{ $sc.stripeUnit | quote }}
+{{- else if $cephCsiDefault.stripeUnit }}
+  stripeUnit: {{ $cephCsiDefault.stripeUnit | quote }}
+{{- end }}
+{{- if $sc.stripeCount }}
+  stripeCount: {{ $sc.stripeCount | quote }}
+{{- else if $cephCsiDefault.stripeCount }}
+  stripeCount: {{ $cephCsiDefault.stripeCount | quote }}
+{{- end }}
+{{- if $sc.objectSize }}
+  objectSize: {{ $sc.objectSize | quote }}
+{{- else if $cephCsiDefault.objectSize }}
+  objectSize: {{ $cephCsiDefault.objectSize | quote }}
+{{- end }}
+  csi.storage.k8s.io/provisioner-secret-name: {{ or $sc.provisionerSecret $defaults.adminSecretName }}
+  csi.storage.k8s.io/provisioner-secret-namespace: {{ or $sc.provisionerSecretNamespace $.Release.Namespace }}
+  csi.storage.k8s.io/controller-expand-secret-name: {{ or $sc.controllerExpandSecret $defaults.adminSecretName }}
+  csi.storage.k8s.io/controller-expand-secret-namespace: {{ or $sc.controllerExpandSecretNamespace $.Release.Namespace }}
+  csi.storage.k8s.io/node-stage-secret-name: {{ or $sc.nodeStageSecret $defaults.adminSecretName }}
+  csi.storage.k8s.io/node-stage-secret-namespace: {{ or $sc.nodeStageSecretNamespace $.Release.Namespace }}
+  csi.storage.k8s.io/fstype: {{ or $sc.fstype $cephCsiDefault.fstype }}
+reclaimPolicy: {{ or $sc.reclaimPolicy $cephCsiDefault.reclaimPolicy }}
+allowVolumeExpansion: {{ or $sc.allowVolumeExpansion $cephCsiDefault.allowVolumeExpansion }}
+{{- if $sc.mountOptions }}
+mountOptions:
+  {{- range $sc.mountOptions }}
+  - {{ . }}
+  {{- end }}
+{{- else if $cephCsiDefault.mountOptions }}
 mountOptions:
-  {{- range .Values.storageClass.mountOptions }}
+  {{- range $cephCsiDefault.mountOptions }}
   - {{ . }}
   {{- end }}
 {{- end }}
+---
+{{- end }}
 {{- end -}}
--
2.25.1
